### PR TITLE
Add documentation for `LogisticRegression` and `SoftmaxRegression`

### PR DESCRIPTION
**Pull request by rcurtin**
Similar to #3560, I added documentation for the logistic regression and softmax regression classifiers.  Compiled documentation as HTML can be seen here:

 * https://www.ratml.org/misc/mlpack-markdown-doc/logistic_regression.html
 * https://www.ratml.org/misc/mlpack-markdown-doc/softmax_regression.html

A number of changes were necessary to make all of these things work (as usual...):

 * Added `MatType` template parameter to `SoftmaxRegression` and used throughout.
 * Modified `LogisticRegression` and `SoftmaxRegression` to hold weights that have the same element type as `MatType`.
 * Added consistent overloads of constructors and `Train()`, which all can optionally take instantiated ensmallen optimizers and callbacks.
 * Added tests for all variants of constructors and `Train()`.
 * Added `Reset()` functionality, which is needed because training is by default incremental.
 * Added reverse-compatible serialization implementations.

All examples were compiled and tested by hand... I would like to build an automated system to do this at some point, but I'll save that for another day.

### CODE REVIEW

**Comment by shrit** [comment]
any reason why the tests are failing ? the docs looks good, I need to go also through the code modification since it seems that there has been some modification and some functions have templates now. However, the failing tests seems to be unrelated

**Comment by shrit** [comment]
some of the links are not working especially related to the examples but I would say this should be resolved before merging right ?

**Comment by rcurtin** [comment]
This one is ready for review too---the builds are passing now (except for the random failure on the Windows build).

**Comment by shrit** [comment]
Okay will look into this one as well tomorrow

File: src/mlpack/tests/softmax_regression_test.cpp [hunk 1]
```diff
@@ -277,37 +283,38 @@ TEST_CASE("SoftmaxRegressionMultipleClasses", "[SoftmaxRegressionTest]")
   GaussianDistribution g4(arma::vec("4.0 1.0 1.0 2.0 7.0"), identity);
   GaussianDistribution g5(arma::vec("1.0 0.0 1.0 8.0 3.0"), identity);
```
Comment (shrit) [inline src/mlpack/tests/softmax_regression_test.cpp:284]: We could have initialised it with the same element type of MatType so we could avoid the conv later but it is fine no need to do anything

**Review by shrit — CHANGES_REQUESTED** [review]
This looks really good, it is amazing, I love these two pages, they are straightforward to the point, have no repetition and are easy to read. I discovered a lot of stuff I did not know at all.

It took me some time to go over it mainly because I wanted to click on each link to be sure that I did not miss anything.

Here are a couple of things that I have noticed:

1. The following links are not working in the two pages (the link to one another), I think they should be working before merging as we discussed above unless if I am missing something
Screenshot 2023-12-15 at 12-59-31 https __www ratml org
Screenshot 2023-12-15 at 12-52-32 https __www ratml org
Screenshot 2023-12-15 at 12-51-12 https __www ratml org
Screenshot 2023-12-15 at 12-50-58 https __www ratml org
Screenshot 2023-12-15 at 12-59-42 https __www ratml org

2. I think the types here mean default type because we can use `float` and other types e,.g., `arma::fmat`. I was confused when I saw it because it seems as if it refers to the _only_ type which is the case.
Maybe change `type` to `default type` or something similar? I know this should be done on all the pages :unamused: but it could be worth it.

Screenshot 2023-12-15 at 13-16-38 https __www ratml org

3. I see we have added tests, and I am a bit surprised they are passing with the fact that we are declaring numbers without specifying the type in some cases. I know the compiler on my machine is translating this to double when I was testing dbscan and knn. I would say would be nice to cast it or define variables to be safe, but you know better than me. 

This could be also the reason we are having a nan here in the failing tests, but it is hard to deduct, 
```
D:\a\1\s\src\mlpack\tests\main_tests\radical_test.cpp(30): FAILED:
due to unexpected exception with message:
  sort(): detected NaN
```
I do not know if Radical test are using logistic regression somewhere

**Comment by rcurtin** [comment]
> The following links are not working in the two pages (the link to one another), I think they should be working before merging as we discussed above unless if I am missing something

I'd actually rather wait to fix these links; I don't know quite how things should be organized overall quite yet, so I was planning to figure that out (hopefully this week), then fix all the links in the individual Markdown files all at once.  Do you think that sounds reasonable?

> I think the types here mean default type because we can use float and other types e,.g., arma::fmat. I was confused when I saw it because it seems as if it refers to the only type which is the case.  Maybe change type to default type or something similar? I know this should be done on all the pages 😒 but it could be worth it.

I spent a long time thinking about it... I wanted to try and keep the documentation simple and use default types where possible.  I could do `default type`, but I would have to denote in the template parameter section what all the types change to.  Another idea would be to note something like this:

```
 * When using a custom `MatType`, the training parameters will change type:
 
 <table goes here>
 
 * The classification parameters will change type too:
  
  <table goes here>
  
  * `lr.Parameters()` will now return a `MatType&`...
  ```
  
That's just a basic idea, but it keeps the different type information constrained to the section where we talk about different `MatType`s.  What do you think?

I also thought about a drop-down selector box that could change types, but that gets really ugly and would mean that the Markdown wasn't easily readable anymore.

> I see we have added tests, and I am a bit surprised they are passing with the fact that we are declaring numbers without specifying the type in some cases. I know the compiler on my machine is translating this to double when I was testing dbscan and knn. I would say would be nice to cast it or define variables to be safe, but you know better than me.

Thanks for pointing this out... I saw the comments you left and I'll address them.  As far as I know, the implicit casting of a double literal to a float isn't an issue, like if we did:

```
float x = 3.0;
```

and I think that's done several times in the tests.  I won't worry about those instances.  I read [this](https://stackoverflow.com/questions/55408191/is-literal-double-to-float-conversion-equal-to-float-literal) on that point.

> This could be also the reason we are having a nan here in the failing tests, but it is hard to deduct,
>
> ```
> D:\a\1\s\src\mlpack\tests\main_tests\radical_test.cpp(30): FAILED:
> due to unexpected exception with message:
>  sort(): detected NaN
> ```
>
> I do not know if Radical test are using logistic regression somewhere

RADICAL doesn't use logistic regression, so I think the failures are unrelated.  I have actually been chasing that failure for a very long time and I've never been able to reproduce it.  I wonder if there is a subtle memory error caused by another test, that then happens to result in the RADICAL test failing... but I am not sure.

File: src/mlpack/methods/logistic_regression/logistic_regression.hpp [hunk 2]
```diff
@@ -37,79 +37,128 @@ template<typename MatType = arma::mat>
...
-   * optimizer type, or use the constructor that takes an instantiated
-   * optimizer.  (This unfortunate situation is a language restriction of C++.)
+   * This constructor uses the default L-BFGS optimizer to train the model.
    *
    * @param predictors Input training variables.
    * @param responses Outputs resulting from input training variables.
    * @param lambda L2-regularization parameter.
    */
+  template<typename... CallbackTypes,
+           typename = typename std::enable_if<IsEnsCallbackTypes<
```
[the remark this answers is not in the document]
Comment (rcurtin) [reply src/mlpack/methods/logistic_regression/logistic_regression.hpp:70]: Mostly I used the `IsEnsCallbackTypes` to avoid function ambiguity... it's not directly relevant for this one, but it is still nice to have.  For some of the `Train()` overloads, you could call e.g. `Train(data, responses, optimizer, 0.3)` or `Train(data, responses, optimizer, ens::TimerStop(0.1))` and there is no way to figure out if you meant that `0.3` should be `lambda` or a callback.  So the `IsEnsCallbackTypes` uses SFINAE to allow the compiler to figure out which overload is desired.

File: src/mlpack/methods/logistic_regression/logistic_regression.hpp [hunk 3]
```diff
@@ -203,8 +360,16 @@ class LogisticRegression
    * @param dataset Set of points to classify.
    * @param probabilities Class probabilities for each point (output).
    */
+  mlpack_deprecated /* to be removed in mlpack 5.0.0 */
```
[the remark this answers is not in the document]
Comment (rcurtin) [reply src/mlpack/methods/logistic_regression/logistic_regression.hpp:363]: https://github.com/mlpack/mlpack/blob/03c0c6a3821bb4e38903d4e1d61ded2a24d444f0/src/mlpack/core/util/deprecated.hpp#L5 :+1:

File: src/mlpack/methods/logistic_regression/logistic_regression_impl.hpp [hunk 4]
```diff
@@ -98,17 +139,50 @@ double LogisticRegression<MatType>::Train(
...
     const
 {
   return size_t(1.0 / (1.0 + std::exp(-parameters(0) - arma::dot(point,
-      parameters.tail_cols(parameters.n_elem - 1)))) +
+      parameters.tail_cols(parameters.n_elem - 1).t()))) +
       (1.0 - decisionBoundary));
 }
 
+template<typename MatType>
+template<typename VecType>
```
[the remark this answers is not in the document]
Comment (rcurtin) [reply src/mlpack/methods/logistic_regression/logistic_regression_impl.hpp:169]: The issue here is, what if I do `lr.Classify(matrix.col(i))`?  Then `MatType` might be `arma::mat`, but `VecType` will be `arma::subview_col` (or similar).  So I have kept the types separate to allow this flexibility... if we made the first parameter `MatType` instead of `VecType`, then calling `lr.Classify(matrix.col(i))` would compile, but the `matrix.col(i)` subexpression would be formed into its own `arma::mat` via copy.

File: src/mlpack/methods/softmax_regression/softmax_regression_impl.hpp [hunk 5]
```diff
@@ -17,118 +17,170 @@
...
+inline SoftmaxRegression<MatType>::SoftmaxRegression(
+    const size_t inputSize,
     const size_t numClasses,
-    const double lambda,
-    const bool fitIntercept,
-    OptimizerType optimizer) :
+    const bool fitIntercept) :
     numClasses(numClasses),
-    lambda(lambda),
+    lambda(0.0001),
```
[the remark this answers is not in the document]
Comment (rcurtin) [reply src/mlpack/methods/softmax_regression/softmax_regression_impl.hpp:26]: `lambda` is always `double` regardless of `MatType` (just for convenience and clarity), so I don't think it's needed here... let me know if you think otherwise. :+1:

File: src/mlpack/tests/logistic_regression_test.cpp [hunk 6]
```diff
@@ -1078,3 +1089,173 @@ TEST_CASE("IncrementalTraining", "[LogisticRegressionTest]")
 
   REQUIRE(acc == Approx(100.0).epsilon(0.03)); // 3% error tolerance.
 }
+
+// Test all constructor variants.  This test is more about checking that all
+// variants compile correctly than anything else.
+TEMPLATE_TEST_CASE("LogisticRegressionAllConstructorsTest",
+    "[LogisticRegressionTest]", arma::mat, arma::fmat)
+{
```
[the remark this answers is not in the document]
Comment (rcurtin) [reply src/mlpack/tests/logistic_regression_test.cpp:1097]: Yeah, as I've written more and more of these documentation PRs, I've started just changing `TEST_CASE`s to `TEMPLATE_TEST_CASE`s... I really like the support.  It gets used super heavily inside of Bandicoot for tests that use all possible types.

File: src/mlpack/tests/softmax_regression_test.cpp (hunk 1 above)
Comment (rcurtin) [reply src/mlpack/tests/softmax_regression_test.cpp:284]: Yeah, the issue here is that `GaussianDistribution` is specific to `arma::mat`, and I didn't want to take the time in this PR to do that too... it would make it even bigger than it already is.  I'll definitely do that before the end of this documentation saga, and will come back here to clean this up then. :+1:

File: src/mlpack/methods/logistic_regression/logistic_regression_function.hpp [hunk 7]
```diff
@@ -128,28 +134,31 @@ class LogisticRegressionFunction
    *    be computed.
    * @param gradient Sparse matrix to output gradient into.
```
[the remark this answers is not in the document]
Comment (rcurtin) [reply src/mlpack/methods/logistic_regression/logistic_regression_function.hpp]: Thanks, fixed in 3100965f56.

File: src/mlpack/methods/logistic_regression/logistic_regression_function_impl.hpp [hunk 8]
```diff
@@ -65,25 +65,28 @@ void LogisticRegressionFunction<MatType>::Shuffle()
...
   //   f(w) = sum(y log(sig(w'x)) + (1 - y) log(sig(1 - w'x))).
   // We want to minimize this function.  L2-regularization is just lambda
   // multiplied by the squared l2-norm of the parameters then divided by two.
+  typedef typename CoordinatesType::elem_type ElemType;
 
   // For the regularization, we ignore the first term, which is the intercept
   // term and take every term except the last one in the decision variable.
-  const double regularization = 0.5 * lambda *
-      arma::dot(parameters.tail_cols(parameters.n_elem - 1),
+  const ElemType regularization = 0.5 * lambda *
```
[the remark this answers is not in the document]
Comment (rcurtin) [reply src/mlpack/methods/logistic_regression/logistic_regression_function_impl.hpp]: 24eeca4caf

File: src/mlpack/methods/logistic_regression/logistic_regression_function_impl.hpp [hunk 9]
```diff
@@ -65,25 +65,28 @@ void LogisticRegressionFunction<MatType>::Shuffle()
...
+      dot(parameters.tail_cols(parameters.n_elem - 1),
       parameters.tail_cols(parameters.n_elem - 1));
 
   // Calculate vectors of sigmoids.  The intercept term is parameters(0, 0) and
   // does not need to be multiplied by any of the predictors.
-  const arma::rowvec sigmoid = 1.0 / (1.0 + arma::exp(-(parameters(0, 0) +
+  const CoordinatesType sigmoid = 1.0 / (1.0 + exp(-(parameters(0, 0) +
```
[the remark this answers is not in the document]
Comment (rcurtin) [reply src/mlpack/methods/logistic_regression/logistic_regression_function_impl.hpp]: Yes, I could probably reason whether or not things will safely work with Armadillo, but Armadillo may change internally... so better to be safe and use your suggestion of `ElemType one`.  I did that for this whole file and also `softmax_regression_function_impl.hpp` (which only had single usages of `1.0`, so I just did it inline), in commit 24eeca4caf.

File: src/mlpack/methods/logistic_regression/logistic_regression_function_impl.hpp [hunk 10]
```diff
@@ -92,9 +95,9 @@ double LogisticRegressionFunction<MatType>::Evaluate(
   // terms for computational efficiency.  Note that the conversion causes some
   // copy and slowdown, but this is so negligible compared to the rest of the
   // calculation it is not worth optimizing for.
-  const double result = arma::accu(arma::log(1.0 -
-      arma::conv_to<arma::rowvec>::from(responses) + sigmoid %
-      (2 * arma::conv_to<arma::rowvec>::from(responses) - 1.0)));
+  const ElemType result = accu(log(1.0 -
+      arma::conv_to<CoordinatesType>::from(responses) + sigmoid %
+      (2 * arma::conv_to<CoordinatesType>::from(responses) - 1.0)));
```
[the remark this answers is not in the document]
Comment (rcurtin) [reply src/mlpack/methods/logistic_regression/logistic_regression_function_impl.hpp]: 24eeca4caf

File: src/mlpack/methods/logistic_regression/logistic_regression_function_impl.hpp [hunk 11]
```diff
@@ -105,74 +108,78 @@ double LogisticRegressionFunction<MatType>::Evaluate(
...
       predictors.cols(begin, begin + batchSize - 1))));
 
   // Compute the objective for the given batch size from a given point.
-  arma::rowvec respD = arma::conv_to<arma::rowvec>::from(responses.subvec(begin,
-      begin + batchSize - 1));
-  const double result = arma::accu(arma::log(1.0 - respD + sigmoid %
-      (2 * respD - 1.0)));
+  CoordinatesType respD = arma::conv_to<CoordinatesType>::from(
+      responses.subvec(begin, begin + batchSize - 1));
+  const ElemType result = accu(log(1.0 - respD + sigmoid % (2 * respD - 1.0)));
```
[the remark this answers is not in the document]
Comment (rcurtin) [reply src/mlpack/methods/logistic_regression/logistic_regression_function_impl.hpp]: 24eeca4caf

**Comment by rcurtin** [comment]
@shrit thank you so much for the comprehensive review!  It is really helpful.  I think I have responded to everything; let me know if not, and let me know what you think of the comments.  I am pretty tired right now so I may have overlooked something (I probably should have just gone to bed, but oh well, too late now :smile:)

**Review by shrit — APPROVED** [review]
Looks good to me

**Comment by shrit** [comment]
Regarding the solution for `default type`, I do not have one, I want something really simple to understand without the need to add more tables or anything.
Maybe if we keep having an example for f32 in the end that would explain that these parameters are not hard coded and are templatized.
Let us keep them this way, maybe we will figure something out in the future for everything

**Comment by rcurtin** [comment]
> Regarding the solution for default type, I do not have one, I want something really simple to understand without the need to add more tables or anything.

I opened https://github.com/mlpack/mlpack/issues/3584 so that the idea doesn't get lost.  I have some ideas, but it may be a little while until I return to it.

**Review by shrit — APPROVED** [review]

**Comment by rcurtin** [comment]
The Python tests will be fixed in #3587, so I will go ahead and merge this despite the failing builds.  (The Go failure appears to be random.)
